### PR TITLE
feat(charts): provide more control over helm tests

### PR DESCRIPTION
**Pull request by schahal**
## Motivation

There are times when we need a bit more control over the `Pod` that's run in `helm test`. For example:

- Update the image to use a different repository (like a pull-thru cache)
- Disable a test entirely due to a finicky CI or local network
- Should a future test image have a breaking change, we can update args/commands via values as a quick workaround rather than waiting for a new helm release

## Changes

This change pulls the defaults into the `values.yaml` for that control.

Optionally, we can enable/disable any of the tests.

## Testing

Ran `helm template test .` before and after the change and confirmed the rendered templates are the same except for:

- adding an explicit `latest` tag for the busybox image
- formatting of command/args list

```diff
--- /tmp/localstack-orig.yaml	2025-03-07 09:35:10
+++ /tmp/localstack-new.yaml	2025-03-07 09:34:16
@@ -494,9 +494,15 @@
 spec:
   containers:
     - name: wget-edge
-      image: busybox
-      command: ['wget']
-      args: ['-q', '--server-response', '--output-document', '-', 'test-localstack:4566/_localstack/health']
+      image: busybox:latest
+      command:
+        - wget
+      args:
+        - "-q"
+        - "--server-response"
+        - "--output-document"
+        - "-"
+        - "test-localstack:4566/_localstack/health"
   restartPolicy: Never
 ---
 # Source: localstack/templates/tests/test-s3.yaml
@@ -517,7 +523,12 @@
   containers:
     - name: awscli-s3
       image: amazon/aws-cli
-      args: ['--debug', '--endpoint-url', 'http://test-localstack:4566', 's3', 'ls']
+      args:
+        - "--debug"
+        - "--endpoint-url"
+        - "http://test-localstack:4566"
+        - "s3"
+        - "ls"
       env:
       - name: AWS_ACCESS_KEY_ID
         value: test
```

### CODE REVIEW

File: charts/localstack/templates/tests/test-connection.yaml [hunk 1]
```diff
@@ -1,3 +1,4 @@
+{{- if .Values.tests.connection.enabled }}
```
Comment (alexrashed) [inline charts/localstack/templates/tests/test-connection.yaml:1]: **question:** This conditional addresses the following point you mentioned in your PR description:

> Disable a test entirely due to a finicky CI or local network

But both these tests are extremely simple. If one of them doesn't work, they will both not work. And in that case, you could just avoid running the tests in the first place, right?
So I don't think this is really necessary for the two tests?

File: charts/localstack/templates/tests/test-connection.yaml [hunk 2]
```diff
@@ -13,8 +14,16 @@ metadata:
     "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
 spec:
   containers:
-    - name: wget-edge
-      image: busybox
-      command: ['wget']
-      args: ['-q', '--server-response', '--output-document', '-', '{{ include "localstack.fullname" . }}:{{ .Values.service.edgeService.targetPort }}/_localstack/health']
+    - name: {{ .Values.tests.connection.name }}
+      image: {{ .Values.tests.connection.image }}
```
Comment (alexrashed) [inline charts/localstack/templates/tests/test-connection.yaml:18]: **question:** Maybe it would make sense to just use one image for all tests? The `aws-cli` image does not contain `wget`, but would contain `curl`.

File: charts/localstack/templates/tests/test-connection.yaml [hunk 3]
```diff
@@ -13,8 +14,16 @@ metadata:
...
+      {{- with .Values.tests.connection.command }}
+      command:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      args:
+        {{- $global := . }}
+        {{- range $arg := index .Values.tests.connection.args }}
+        - {{ tpl $arg $global | quote }}
+        {{- end }}
```
Comment (alexrashed) [inline charts/localstack/templates/tests/test-connection.yaml:27]: **question:** This effectively extracts all the content of the test to the values. In your PR description you mentioned the motivation for that:

> Should a future test image have a breaking change, we can update args/commands via values as a quick workaround rather than waiting for a new helm release

I don't really see that. These tests are fundamental to LocalStack and won't break, not even with any major releases / other breaking changes along the way.
Do you already see the need to adjust the commands here in your scenario?

**Review by alexrashed — COMMENTED** [review]
Thanks a lot @schahal for the contributions! This PR actually fixes #138 which was just raised by @thwint! 🥳 
I added a few comments with follow up questions, because I feel like this is generalizing quite a lot which might not be necessary?

File: charts/localstack/templates/tests/test-connection.yaml (hunk 1 above)
Comment (schahal) [reply charts/localstack/templates/tests/test-connection.yaml:1]: Great question! I wanted to avoid technicals (because I agree these are very basic tests), but here's a CI / Cluster (Kuberentes-in-Docker [KIND]) weird networking hiccup which prompted me to want this feature:

Our CI tool runs `helm install --wait` followed by `helm test ...`

This immediate jump from `install` to `test` sometimes (50/50) either causes test 1 to fail _**or**_ test 2 fail

<details>
<summary><b>Click here</b> for an example with one test passing but the other failing</summary>  

```
$ helm install foo --wait .; helm test foo
NAME: foo
LAST DEPLOYED: Thu Mar 13 09:34:01 2025
NAMESPACE: observability
STATUS: deployed
REVISION: 1
NOTES:
1. Get the application URL by running these commands:
  export NODE_PORT=$(kubectl get --namespace "observability" -o jsonpath="{.spec.ports[0].nodePort}" services foo-localstack)
  export NODE_IP=$(kubectl get nodes --namespace "observability" -o jsonpath="{.items[0].status.addresses[0].address}")
  echo http://$NODE_IP:$NODE_PORT
NAME: foo
LAST DEPLOYED: Thu Mar 13 09:34:01 2025
NAMESPACE: observability
STATUS: deployed
REVISION: 1
TEST SUITE:     foo-localstack-test-connection
Last Started:   Thu Mar 13 09:34:33 2025
Last Completed: Thu Mar 13 09:34:37 2025
Phase:          Succeeded
TEST SUITE:     foo-localstack-test-s3
Last Started:   Thu Mar 13 09:34:37 2025
Last Completed: Thu Mar 13 09:34:42 2025
Phase:          Failed
NOTES:
1. Get the application URL by running these commands:
  export NODE_PORT=$(kubectl get --namespace "observability" -o jsonpath="{.spec.ports[0].nodePort}" services foo-localstack)
  export NODE_IP=$(kubectl get nodes --namespace "observability" -o jsonpath="{.items[0].status.addresses[0].address}")
  echo http://$NODE_IP:$NODE_PORT
Error: 1 error occurred:
	* pod foo-localstack-test-s3 failed

$ kubectl logs foo-localstack-test-s3
[...]
botocore.exceptions.EndpointConnectionError: Could not connect to the endpoint URL: "http://foo-localstack:4566/"
```
</details>

--- BREAK ---

This can be mitigated by:
- Either adding a `sleep 5` in between... e.g., ` helm install foo --wait .; sleep 5; helm test foo` (this is untenable in large CI environments with hundreds of tests the CI tool is a "global library"
- (or) Disabling the test altogether (because I have other tests that connect to ci/local localstack fine, so these tests are redundant

Having the option to disable either test and/or update command to add e.g., sleep just for localstack give users the ability to make slight tweaks based on their environments (right now, I have to remove the entire CI step completely from the couple jobs that use localstack).

I believe in 99.99% situations no one will be changing these except for situations like above (I'll open an issue in https://github.com/kubernetes-sigs/kind/ to see if there's some (un)known cni issue there regarding this)

File: charts/localstack/templates/tests/test-connection.yaml (hunk 3 above)
Comment (schahal) [reply charts/localstack/templates/tests/test-connection.yaml:27]: > Do you already see the need to adjust the commands here in your scenario?

Wanna refer to the "sleep" situation in the [above comment thread](https://github.com/localstack/helm-charts/pull/137#discussion_r1993965123) (i put some technical details in that thread)?... it's where adding a `sleep` can help mitigate network setup delays

File: charts/localstack/templates/tests/test-connection.yaml (hunk 1 above)
Comment (alexrashed) [reply charts/localstack/templates/tests/test-connection.yaml:1]: Thanks for the detailed explanation! But in my opinion it would really be great to enhance the test it self. If this is the case, we should just make the tests more robust, for example with retries. This way everyone benefits from the changes. If there is an actual issue the tests should not be successful.

File: charts/localstack/templates/tests/test-connection.yaml (hunk 3 above)
Comment (alexrashed) [reply charts/localstack/templates/tests/test-connection.yaml:27]: As mentioned above, I really don't think this is the way forward. If you are seeing issues with the tests in your environment, we need to adjust the tests to be more robust. If the test does not work in your environment, we should fix the test, or you need to find another way to test the chart.

File: charts/localstack/templates/tests/test-connection.yaml (hunk 2 above)
Comment (alexrashed) [reply charts/localstack/templates/tests/test-connection.yaml:18]: That was really just a suggestion to minimize the amount of necessary values to an absolute minimum. If you replace the curl with a functional equivalent, you could get the test image config in a single value. Another alternative would be an image prefix applied to all images in the repo.
In this case you would have to configure just the prefix with your private repo. But this would mean that the images need to be in the repo with the same name (but with the possibility of a shared prefix).

File: charts/localstack/templates/tests/test-connection.yaml (hunk 3 above)
Comment (schahal) [reply charts/localstack/templates/tests/test-connection.yaml:27]: > If the test does not work in your environment, we should fix the test, or you need to find another way to test the chart.

That's the intention of having users control the test command/args (if needed) for their environment.

Are you suggesting we use this opportunity to re-write the tests for added robustness and keep the tests hard coded in the template? 

If so, I can certainly play around with some valid new tests sometime next week and update the branch.

Comment (alexrashed) [reply charts/localstack/templates/tests/test-connection.yaml:27]: Yes, exactly. If the test is flaky in your environment, we should fix the test 😉

Comment (schahal) [reply charts/localstack/templates/tests/test-connection.yaml]: Added in https://github.com/localstack/helm-charts/pull/137/commits/b6a26448c2094cc34994ba344b264efef859b346

File: charts/localstack/templates/tests/test-connection.yaml (hunk 2 above)
Comment (schahal) [reply charts/localstack/templates/tests/test-connection.yaml]: Switched to `amazon/aws-cli` for both... kept localstack (and general consensus) convention of `image.repository` and `image.tag`: https://github.com/localstack/helm-charts/pull/137/commits/b6a26448c2094cc34994ba344b264efef859b346

File: charts/localstack/templates/tests/test-connection.yaml (hunk 1 above)
Comment (schahal) [reply charts/localstack/templates/tests/test-connection.yaml]: Enhanced in https://github.com/localstack/helm-charts/pull/137/commits/b6a26448c2094cc34994ba344b264efef859b346

**Comment by schahal** [comment]
@alexrashed re-review / approve workflow please?

Anyone else following: the original tenant/description of this PR to keep defaults the same has changed (per discussions above)... we're changing the tests to be a bit more robust:

## Summary of Changes

### Single Test Image

- Use single test image (`amazon/aws-cli`): same conventions as everything else to expose `repository` and `tag`, if necessary

### Connection Test

- Moved from `wget` to `curl` --> this image's curl strips aways the `--retry` option, so we have to add our own simple retry mechanism

### AWS CLI `s3 ls` Test

- For the `aws s3` command, AWS CLI V2 doesn't support the `AWS_MAX_ATTEMPTS` (https://github.com/aws/aws-cli/issues/5653), so we use our own simple retry mechanism below.

☝️ Both have `RETRIES=5` and `DELAY=2` in between for a worst case 10 seconds of retries.

<details>
<summary><b>Test Result in my Spotty Environment</b></summary>

```
helm install foo --wait .; helm test foo
NAME: foo
LAST DEPLOYED: Fri Mar 14 15:41:20 2025
NAMESPACE: default
STATUS: deployed
REVISION: 1
NOTES:
1. Get the application URL by running these commands:
  export NODE_PORT=$(kubectl get --namespace "default" -o jsonpath="{.spec.ports[0].nodePort}" services foo-localstack)
  export NODE_IP=$(kubectl get nodes --namespace "default" -o jsonpath="{.items[0].status.addresses[0].address}")
  echo http://$NODE_IP:$NODE_PORT
NAME: foo
LAST DEPLOYED: Fri Mar 14 15:41:20 2025
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE:     foo-localstack-test-connection
Last Started:   Fri Mar 14 15:41:42 2025
Last Completed: Fri Mar 14 15:41:45 2025
Phase:          Succeeded
TEST SUITE:     foo-localstack-test-s3
Last Started:   Fri Mar 14 15:41:45 2025
Last Completed: Fri Mar 14 15:41:49 2025
Phase:          Succeeded
NOTES:
1. Get the application URL by running these commands:
  export NODE_PORT=$(kubectl get --namespace "default" -o jsonpath="{.spec.ports[0].nodePort}" services foo-localstack)
  export NODE_IP=$(kubectl get nodes --namespace "default" -o jsonpath="{.items[0].status.addresses[0].address}")
  echo http://$NODE_IP:$NODE_PORT
```
</details>

☝️ I've also tested the negative of this by hardcoding an unreachable URL and validating a helpful failure

I've kept `tests.enabled` (from initial commit) because I use localstack as a subchart with my own tests that use it and these tests are redundant.

**Review by alexrashed — APPROVED** [review]
Thanks @schahal for the contribution and for following up on the discussion / comments on the review! The changes are looking great! I'll merge this PR right away and create a new release 🥳  📦 🚀
